File: src/MyServiceBus.AzureServiceBus/AzureServiceBusFactoryConfigurator.cs
using Azure.Messaging.ServiceBus;
using Microsoft.Extensions.DependencyInjection;
using MyServiceBus.Topology;

namespace MyServiceBus;

public sealed class AzureServiceBusFactoryConfigurator : IAzureServiceBusFactoryConfigurator, IBusFactoryConfigurator
{
    internal const string EmulatorConnectionString =
        "Endpoint=sb://localhost;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SAS_KEY_VALUE;UseDevelopmentEmulator=true;";

    private readonly Dictionary<Type, string> _entityNames = new();
    private readonly List<Action<IMessageBus, IServiceProvider>> _endpointActions = new();
    private Type _consumerFactoryType = typeof(DefaultConstructorConsumerFactory<>);

    public string ConnectionString { get; private set; } = EmulatorConnectionString;

    public string? ManagementConnectionString { get; private set; }

    public AzureServiceBusTopologyMode TopologyMode { get; private set; } = AzureServiceBusTopologyMode.Create;

    public int PrefetchCount { get; private set; }

    public IEndpointNameFormatter? EndpointNameFormatter { get; private set; }

    public IMessageEntityNameFormatter? EntityNameFormatter { get; private set; } =
        AzureServiceBusMessageEntityNameFormatter.Instance;

    public Func<string, string> TemporaryEndpointNameFormatter { get; private set; } = name => name;

    public void Host(string connectionString)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(connectionString);
        _ = ServiceBusConnectionStringProperties.Parse(connectionString);
        ConnectionString = connectionString;
    }

    public void ManagementEndpoint(string connectionString)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(connectionString);
        _ = ServiceBusConnectionStringProperties.Parse(connectionString);
        ManagementConnectionString = connectionString;
    }

    public void UsePreProvisionedTopology() => TopologyMode = AzureServiceBusTopologyMode.PreProvisioned;

    public void SetPrefetchCount(int prefetchCount)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(prefetchCount);
        PrefetchCount = prefetchCount;
    }

    public void SetEndpointNameFormatter(IEndpointNameFormatter formatter)
    {
        ArgumentNullException.ThrowIfNull(formatter);
        EndpointNameFormatter = formatter;
    }

    public void SetEntityNameFormatter(IMessageEntityNameFormatter formatter)
    {
        ArgumentNullException.ThrowIfNull(formatter);
        EntityNameFormatter = formatter;
    }

    public void SetTemporaryEndpointNameFormatter(Func<string, string> formatter)
    {
        ArgumentNullException.ThrowIfNull(formatter);
        TemporaryEndpointNameFormatter = name =>
        {
            var formatted = formatter(name);
            return string.IsNullOrWhiteSpace(formatted)
                ? throw new InvalidOperationException("The temporary endpoint name formatter returned a blank name.")
                : formatted;
        };
    }

    public void Message<T>(Action<AzureServiceBusMessageConfigurator> configure)
    {
        ArgumentNullException.ThrowIfNull(configure);
        configure(new AzureServiceBusMessageConfigurator(typeof(T), _entityNames));
    }

    public string GetEntityName(Type messageType)
    {
        ArgumentNullException.ThrowIfNull(messageType);
        return _entityNames.TryGetValue(messageType, out var configuredName)
            ? configuredName
            : EntityNameFormatter!.FormatEntityName(messageType);
    }

    public void ReceiveEndpoint(string queueName, Action<AzureServiceBusReceiveEndpointConfigurator> configure)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(queueName);
        ArgumentNullException.ThrowIfNull(configure);
        configure(new AzureServiceBusReceiveEndpointConfigurator(queueName, GetEntityName, _endpointActions));
    }

    public void SetConsumerFactory(Type consumerFactoryType)
    {
        ArgumentNullException.ThrowIfNull(consumerFactoryType);
        _consumerFactoryType = consumerFactoryType;
    }

    public void ConfigureEndpoints(IBusRegistrationContext context)
    {
        ArgumentNullException.ThrowIfNull(context);
        var registry = context.ServiceProvider.GetRequiredService<TopologyRegistry>();

        foreach (var consumer in registry.Consumers)
        {
            var messageType = consumer.Bindings.First().MessageType;
            var queueName = EndpointNameFormatter?.Format(messageType) ?? consumer.QueueName;
            var consumerType = consumer.ConsumerType;
            ReceiveEndpoint(queueName, endpoint =>
            {
                var method = typeof(AzureServiceBusReceiveEndpointConfigurator)
                    .GetMethod(nameof(AzureServiceBusReceiveEndpointConfigurator.ConfigureConsumer))!
                    .MakeGenericMethod(consumerType);
                method.Invoke(endpoint, [context]);
            });
        }
    }

    internal void Apply(IMessageBus bus, IServiceProvider provider)
    {
        foreach (var action in _endpointActions)
            action(bus, provider);
    }

    public IMessageBus Build()
    {
        var services = new ServiceCollection();
        Configure(services);
        var provider = services.BuildServiceProvider();
        foreach (var action in provider.GetServices<IPostBuildAction>())
            action.Execute(provider);
        return provider.GetRequiredService<IMessageBus>();
    }

    public void Configure(IServiceCollection services)
    {
        var configurator = new BusRegistrationConfigurator(services);
        configurator.Build();
        services.AddSingleton(typeof(IConsumerFactory<>), _consumerFactoryType);
        AzureServiceBusServiceCollectionExtensions.RegisterTransport(
            services,
            this,
            (context, cfg) => ((AzureServiceBusFactoryConfigurator)cfg).ConfigureEndpoints(context));
    }
}

File: src/Java/myservicebus-azure-service-bus/src/main/java/com/myservicebus/azure/servicebus/AzureServiceBusFactoryConfigurator.java
package com.myservicebus.azure.servicebus;

import com.myservicebus.BusFactoryConfigurator;
import com.myservicebus.BusRegistrationConfiguratorImpl;
import com.myservicebus.BusRegistrationContext;
import com.myservicebus.ConsumerFactory;
import com.myservicebus.DefaultConstructorConsumerFactory;
import com.myservicebus.EndpointNameFormatter;
import com.myservicebus.MessageBus;
import com.myservicebus.MessageBusImpl;
import com.myservicebus.MessageEntityNameFormatter;
import com.myservicebus.di.ServiceCollection;
import com.myservicebus.di.ServiceProvider;
import com.myservicebus.serialization.MessageSerializer;
import com.myservicebus.topology.ConsumerTopology;
import com.myservicebus.topology.TopologyRegistry;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.function.Function;

public final class AzureServiceBusFactoryConfigurator implements BusFactoryConfigurator {
    public static final String EMULATOR_CONNECTION_STRING =
            "Endpoint=sb://localhost;SharedAccessKeyName=RootManageSharedAccessKey;"
                    + "SharedAccessKey=SAS_KEY_VALUE;UseDevelopmentEmulator=true;";

    private String connectionString = EMULATOR_CONNECTION_STRING;
    private String managementConnectionString;
    private AzureServiceBusTopologyMode topologyMode = AzureServiceBusTopologyMode.CREATE;
    private int prefetchCount;
    private Function<String, String> temporaryEndpointNameFormatter = name -> name;
    private EndpointNameFormatter endpointNameFormatter;
    private MessageEntityNameFormatter entityNameFormatter = AzureServiceBusMessageEntityNameFormatter.INSTANCE;
    private final Map<Class<?>, String> entityNames = new HashMap<>();
    private final List<AzureServiceBusReceiveEndpointConfigurator.HandlerRegistration<?>> handlers =
            new ArrayList<>();
    private java.util.function.BiFunction<ServiceProvider, Class<?>, ConsumerFactory> consumerFactory =
            (provider, type) -> new DefaultConstructorConsumerFactory();

    public void host(String value) {
        endpoint(value);
        connectionString = value;
    }

    public void managementEndpoint(String value) {
        endpoint(value);
        managementConnectionString = value;
    }

    public void usePreProvisionedTopology() {
        topologyMode = AzureServiceBusTopologyMode.PRE_PROVISIONED;
    }

    public void setPrefetchCount(int value) {
        if (value < 0) {
            throw new IllegalArgumentException("Azure Service Bus prefetch count cannot be negative");
        }
        prefetchCount = value;
    }

    public void setEndpointNameFormatter(EndpointNameFormatter value) {
        endpointNameFormatter = value;
    }

    public void setEntityNameFormatter(MessageEntityNameFormatter value) {
        if (value == null) {
            throw new IllegalArgumentException("Entity name formatter cannot be null");
        }
        entityNameFormatter = value;
    }

    public void setTemporaryEndpointNameFormatter(Function<String, String> value) {
        if (value == null) {
            throw new IllegalArgumentException("Temporary endpoint name formatter cannot be null");
        }
        temporaryEndpointNameFormatter = name -> {
            String formatted = value.apply(name);
            if (formatted == null || formatted.isBlank()) {
                throw new IllegalStateException("The temporary endpoint name formatter returned a blank name");
            }
            return formatted;
        };
    }

    public void setConsumerFactory(
            java.util.function.BiFunction<ServiceProvider, Class<?>, ConsumerFactory> value) {
        consumerFactory = value;
    }

    public <T> void message(Class<T> messageType, Consumer<AzureServiceBusMessageConfigurator<T>> configure) {
        configure.accept(new AzureServiceBusMessageConfigurator<>(messageType, entityNames));
    }

    public String getEntityName(Class<?> messageType) {
        return entityNames.getOrDefault(messageType, entityNameFormatter.formatEntityName(messageType));
    }

    public void receiveEndpoint(
            String queueName,
            Consumer<AzureServiceBusReceiveEndpointConfigurator> configure) {
        configure.accept(new AzureServiceBusReceiveEndpointConfigurator(queueName, this::getEntityName, handlers));
    }

    public void configureEndpoints(BusRegistrationContext context) {
        TopologyRegistry registry = context.getServiceProvider().getService(TopologyRegistry.class);
        for (ConsumerTopology definition : registry.getConsumers()) {
            String queueName = definition.resolveEndpointName(endpointNameFormatter);
            receiveEndpoint(queueName, endpoint -> endpoint.configureConsumer(context, definition));
        }
    }

    void applyHandlers(MessageBusImpl bus) throws Exception {
        for (AzureServiceBusReceiveEndpointConfigurator.HandlerRegistration<?> handler : handlers) {
            applyHandler(bus, handler);
        }
    }

    @SuppressWarnings({ "rawtypes", "unchecked" })
    private static void applyHandler(
            MessageBusImpl bus,
            AzureServiceBusReceiveEndpointConfigurator.HandlerRegistration handler) throws Exception {
        MessageSerializer serializer = handler.serializerClass() != null
                ? (MessageSerializer) handler.serializerClass().getDeclaredConstructor().newInstance()
                : null;
        bus.addHandler(
                handler.queueName(),
                handler.messageType(),
                handler.entityName(),
                handler.handler(),
                handler.retryCount(),
                handler.retryDelay(),
                handler.prefetchCount(),
                null,
                serializer);
    }

    public String getConnectionString() {
        return connectionString;
    }

    public String getManagementConnectionString() {
        return managementConnectionString;
    }

    public AzureServiceBusTopologyMode getTopologyMode() {
        return topologyMode;
    }

    public int getPrefetchCount() {
        return prefetchCount;
    }

    public Function<String, String> getTemporaryEndpointNameFormatter() {
        return temporaryEndpointNameFormatter;
    }

    @Override
    public MessageBus build() {
        ServiceCollection services = ServiceCollection.create();
        configure(services);
        ServiceProvider provider = services.buildServiceProvider();
        return provider.getService(MessageBus.class);
    }

    @Override
    public void configure(ServiceCollection services) {
        BusRegistrationConfiguratorImpl configurator = new BusRegistrationConfiguratorImpl(services);
        AzureServiceBusTransport.configure(configurator, this);
        configurator.complete();
        services.addSingleton(MessageBus.class, provider -> () -> {
            BusRegistrationContext context = new BusRegistrationContext(provider);
            configureEndpoints(context);
            MessageBusImpl bus = new MessageBusImpl(provider, type -> consumerFactory.apply(provider, type));
            try {
                applyHandlers(bus);
            } catch (Exception exception) {
                throw new RuntimeException("Failed to apply Azure Service Bus handlers", exception);
            }
            return bus;
        });
    }

    private static void endpoint(String connectionString) {
        if (connectionString == null || connectionString.isBlank()) {
            throw new IllegalArgumentException("Azure Service Bus connection string cannot be blank");
        }
        AzureServiceBusTransportFactory.endpoint(connectionString);
    }
}
